File: src/agent_nexus/models/runtime.py
"""Python Runtime models: Variable, Function, Type, ExecutionResult, SecurityViolation."""

from __future__ import annotations

from typing import Any

from pydantic import Field

from agent_nexus.models._common import FrozenModel


class Variable(FrozenModel):
    """A named Python object held in the Runtime namespace.

    Variables persist across execution rounds within a single Agent process.
    The actual Python value is not serialized; only metadata is stored here.
    """

    name: str = Field(min_length=1)
    description: str = ""
    value: Any = None
    type_name: str | None = None


class Function(FrozenModel):
    """A callable Python function registered in the Runtime namespace.

    The callable itself cannot be serialized; only metadata is stored.
    """

    name: str = Field(min_length=1)
    description: str = ""
    signature: str | None = None
    is_async: bool = False


class RuntimeType(FrozenModel):
    """A Python type (class) registered in the Runtime namespace.

    Includes an optional JSON Schema for LLM context injection.
    Named `RuntimeType` to avoid shadowing Python's builtin `type`.
    """

    name: str = Field(min_length=1)
    description: str = ""
    python_type: str | None = None
    json_schema: dict[str, Any] | None = None


class ExecutionResult(FrozenModel):
    """Result of executing Python code in the Runtime."""

    success: bool
    output: str = ""
    error: str | None = None
    variables_created: list[str] = Field(default_factory=list)


class SecurityViolation(FrozenModel):
    """A single security rule violation detected by SecurityChecker (AST-level)."""

    rule_type: str = Field(min_length=1)
    node_type: str = Field(min_length=1)
    code_snippet: str = ""
    message: str = ""

File: crates/ap-core/src/models/runtime.rs
//! Python Runtime models: Variable, Function, RuntimeType, ExecutionResult, SecurityViolation.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Variable {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<serde_json::Value>,
    #[serde(default)]
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub type_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Function {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signature: Option<String>,
    #[serde(default)]
    pub is_async: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RuntimeType {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub python_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub json_schema: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ExecutionResult {
    pub success: bool,
    #[serde(default)]
    pub output: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(default)]
    pub variables_created: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SecurityViolation {
    pub rule_type: String,
    pub node_type: String,
    #[serde(default)]
    pub code_snippet: String,
    #[serde(default)]
    pub message: String,
}
